### PR TITLE
Email Seating Assignments (Optimized) 

### LINKED ISSUE #3026: Send personalized emails from submitty with test seating assignments
(combine this with discussion forum announcement notification?  Or separate?)

### PR DESCRIPTION
**Pull request by kasra-sheik**
This PR corrects some performance issue related to #3325. 

closes #3026

### CODE REVIEW

**Comment by bmcutler** [comment]
@kasra-sheik there are merge conflicts, so I can't merge this.  Please fix :)  Thanks

**Comment by MasterOdin** [comment]
Before we can review this, you'll have to perform some amount of resolution of the merge conflicts resulting from 86cea3d3441419089b1a3058d01c663e6807294c. My suggestion would be to do something like:
```
git checkout master
git pull
git checkout email_seating_assignments
git pull
git merge master
git revert 86cea3d3441419089b1a3058d01c663e6807294c
git cherry-pick 5bd610c^..9c43509
git checkout -b <branch_name>
git push
```

You could replace the last two steps with:
```
git branch -D email_seating_assignments_optimized
git checkout -b email_seating_assignments_optimized
git push --set-upstream origin email_seating_assignments_optimized --force
```
to overwrite the branch on GH and update this PR.

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+ public function __construct(Core $core) {
+        parent::__construct($core);
+    }
+
+	public function run() {
+
+		switch($_REQUEST['action']) {
+				case 'send_email':
+						$this->emailSeatingAssignments();
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: Put a break here.

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			$email_data = [
+						"subject" => $this->replaceSeatingAssignmentMessagePlaceholders($seating_assignment_subject, $room_seating_json),
+						"body" => $this->replaceSeatingAssignmentMessagePlaceholders($seating_assignment_body, $room_seating_json)
+				];
+
+				$this->core->getQueries()->createEmail($email_data, $user_email);
+		}
+
+			$this->core->addSuccessMessage("Seating assignments have been sucessfully emailed!");
+			return $this->core->redirect($this->core->buildUrl());
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: Make the indentation consistent here

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		}
+		if(array_key_exists("zone", $seatingAssignmentData)){
+			$seatingAssignmentMessage = str_replace('{$exam_zone}', $seatingAssignmentData["zone"], $seatingAssignmentMessage);
+		}
+		if(array_key_exists("row", $seatingAssignmentData)){
+			$seatingAssignmentMessage = str_replace('{$exam_row}', $seatingAssignmentData["row"], $seatingAssignmentMessage);
+		}
+		if(array_key_exists("seat", $seatingAssignmentData)){
+			$seatingAssignmentMessage = str_replace('{$exam_seat}', $seatingAssignmentData["seat"], $seatingAssignmentMessage);
+		}
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: You could rewrite this to be more simply:
```
$replaces = [
    'gradeable' => 'gradeable_id',
    'date' => 'exam_date',
    'time' => 'exam_time',
    'building' => 'exam_building',
    'room' => 'exam_room',
    'zone' => 'exam_zone',
    'row' => 'exam_row',
    'seat' => 'exam_seat',
 ];
foreach ($replaces as $key => $variable) {
    if (isset($seatingAssignmentData[$key])) {
        $seatingAssignmentMessage = str_replace('{$' . $variable . '}', $seatingAssignmentData[$key], $seatingAssignmentMessage);
    }
}
```

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+				];
+
+				$this->core->getQueries()->createEmail($email_data, $user_email);
+		}
+
+			$this->core->addSuccessMessage("Seating assignments have been sucessfully emailed!");
+			return $this->core->redirect($this->core->buildUrl());
+	}
+
+	private function replaceSeatingAssignmentMessagePlaceholders($seatingAssignmentMessage, $seatingAssignmentData) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: change the function definition to be `replacePlaceholders($string, $data)`. We don't need that level of verbosity in the function/variables.

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -3773,20 +3773,40 @@ public function getHasSubmission(gradeable\Gradeable $gradeable, Submitter $subm
...
 
     /**
-     * Gets a list of emails for all active particpants in a course  
+      * Retruns true if a given student has a null registration for a course_email
+      *@param $user_id associated user id of the given user
+      */
+    public function hasDroppedCourse($user_id){
+      $this->course_db->query('SELECT registration_section from users WHERE user_id=?', [$user_id]);
+      return $this->course_db->row()['registration_section'] == null;
+    }
```
Comment (MasterOdin) [inline site/app/libraries/database/DatabaseQueries.php]: Do we still need this function? It doesn't look like it's used.

File: site/app/views/admin/EmailRoomSeatingView.php [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace app\views\admin;
+
+use app\views\AbstractView;
+//TODO: add this as a general view in the future..
```
Comment (MasterOdin) [inline site/app/views/admin/EmailRoomSeatingView.php]: what is this "TODO" supposed to mean? Try and rewrite this with the idea that someone not you might look at this and follow through on it.

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
+<?php
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: Use four spaces for indentation, not tabs.

**Comment by bmcutler** [comment]
@kasra-sheik please address Matt's comments and resolve conflicts :)

File: site/app/controllers/admin/EmailRoomSeatingController.php [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            'zone' => 'exam_zone',
+            'row' => 'exam_row',
+            'seat' => 'exam_seat',
+        ];   
+
+        foreach($replaces as $key => $variable) {
+        	if(isset($data[$key])) {
+        		$message = str_replace('{$' . $variable . '}', $data[$key], $message); 
+        	}
+        }
```
Comment (MasterOdin) [inline site/app/controllers/admin/EmailRoomSeatingController.php]: I should have been more clear as you can do the above for most of the variables you had except for the course_name:
```
$message = str_replace('{$course_name}', $this->core->getConfig()->getCourse(), $message);
```

File: site/app/templates/admin/Configuration.twig [hunk 9]
```diff
@@ -224,7 +224,9 @@
                         </option>
                     {% endfor %}
                 </select>
-                 <a href="{{ email_room_seating_url }}" class="btn btn-primary" id="email_seating_assignment" style="margin-top: 25px">Email Seating Assignments</a>
+                <br>
```
Comment (MasterOdin) [inline site/app/templates/admin/Configuration.twig]: We probably want this a tag somewhere (probably under the new instructor only feature) to actually send the emails?

**Pull request merged**